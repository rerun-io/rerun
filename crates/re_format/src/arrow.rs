//! Formatting for tables of Arrow arrays

use std::fmt::Formatter;

use arrow2::{
    array::{get_display, Array, ListArray},
    datatypes::{DataType, IntervalUnit, TimeUnit},
};
use comfy_table::{presets, Cell, Table};

use re_tuid::{external::re_types_core::Loggable as _, Tuid};

// ---

// TODO(#1775): Registering custom formatters should be done from other crates:
// A) Because `re_format` cannot depend on other crates (cyclic deps)
// B) Because how to deserialize and inspect some type is a private implementation detail of that
//    type, re_format shouldn't know how to deserialize a TUID…

type CustomFormatter<'a, F> = Box<dyn Fn(&mut F, usize) -> std::fmt::Result + 'a>;

pub fn get_custom_display<'a, F: std::fmt::Write + 'a>(
    _column_name: &'a str,
    array: &'a dyn Array,
    null: &'static str,
) -> CustomFormatter<'a, F> {
    // NOTE: If the top-level array is a list, it's probably not the type we're looking for: we're
    // interested in the type of the array that's underneath.
    let datatype = (|| match array.data_type().to_logical_type() {
        DataType::List(_) => array
            .as_any()
            .downcast_ref::<ListArray<i32>>()?
            .iter()
            .next()?
            .map(|array| array.data_type().clone()),
        _ => Some(array.data_type().clone()),
    })();

    if let Some(DataType::Extension(name, _, _)) = datatype {
        // TODO(#1775): This should be registered dynamically.
        if name.as_str() == Tuid::name() {
            return Box::new(|w, index| {
                if let Some(tuid) = parse_tuid(array, index) {
                    w.write_fmt(format_args!("{tuid}"))
                } else {
                    w.write_str("<ERR>")
                }
            });
        }
    }

    get_display(array, null)
}

// TODO(#1775): This should be defined and registered by the `re_tuid` crate.
fn parse_tuid(array: &dyn Array, index: usize) -> Option<Tuid> {
    let (array, index) = match array.data_type().to_logical_type() {
        // Legacy MsgId lists: just grab the first value, they're all identical
        DataType::List(_) => (
            array
                .as_any()
                .downcast_ref::<ListArray<i32>>()?
                .value(index),
            0,
        ),
        // New control columns: it's not a list to begin with!
        _ => (array.to_boxed(), index),
    };

    let tuids = Tuid::from_arrow(array.as_ref()).ok()?;
    tuids.get(index).copied()
}

// ---

//TODO(john) move this and the Display impl upstream into arrow2
#[repr(transparent)]
pub struct DisplayTimeUnit(TimeUnit);

impl std::fmt::Display for DisplayTimeUnit {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let s = match self.0 {
            TimeUnit::Second => "s",
            TimeUnit::Millisecond => "ms",
            TimeUnit::Microsecond => "us",
            TimeUnit::Nanosecond => "ns",
        };
        f.write_str(s)
    }
}

//TODO(john) move this and the Display impl upstream into arrow2
#[repr(transparent)]
pub struct DisplayIntervalUnit(IntervalUnit);

impl std::fmt::Display for DisplayIntervalUnit {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let s = match self.0 {
            IntervalUnit::YearMonth => "year/month",
            IntervalUnit::DayTime => "day/time",
            IntervalUnit::MonthDayNano => "month/day/nano",
        };
        f.write_str(s)
    }
}

//TODO(john) move this and the Display impl upstream into arrow2
#[repr(transparent)]
pub struct DisplayDataType(DataType);

impl std::fmt::Display for DisplayDataType {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let s = match &self.0 {
            DataType::Null => "null",
            DataType::Boolean => "bool",
            DataType::Int8 => "i8",
            DataType::Int16 => "i16",
            DataType::Int32 => "i32",
            DataType::Int64 => "i64",
            DataType::UInt8 => "u8",
            DataType::UInt16 => "u16",
            DataType::UInt32 => "u32",
            DataType::UInt64 => "u64",
            DataType::Float16 => "f16",
            DataType::Float32 => "f32",
            DataType::Float64 => "f64",
            DataType::Timestamp(unit, timezone) => {
                let s = if let Some(tz) = timezone {
                    format!("timestamp({}, {tz})", DisplayTimeUnit(*unit))
                } else {
                    format!("timestamp({})", DisplayTimeUnit(*unit))
                };
                return f.write_str(&s);
            }
            DataType::Date32 => "date32",
            DataType::Date64 => "date64",
            DataType::Time32(unit) => {
                let s = format!("time32({})", DisplayTimeUnit(*unit));
                return f.write_str(&s);
            }
            DataType::Time64(unit) => {
                let s = format!("time64({})", DisplayTimeUnit(*unit));
                return f.write_str(&s);
            }
            DataType::Duration(unit) => {
                let s = format!("duration({})", DisplayTimeUnit(*unit));
                return f.write_str(&s);
            }
            DataType::Interval(unit) => {
                let s = format!("interval({})", DisplayIntervalUnit(*unit));
                return f.write_str(&s);
            }
            DataType::Binary => "bin",
            DataType::FixedSizeBinary(size) => return write!(f, "fixed-bin[{size}]"),
            DataType::LargeBinary => "large-bin",
            DataType::Utf8 => "str",
            DataType::LargeUtf8 => "large-string",
            DataType::List(ref field) => {
                let s = format!("list[{}]", DisplayDataType(field.data_type().clone()));
                return f.write_str(&s);
            }
            DataType::FixedSizeList(field, len) => {
                let s = format!(
                    "fixed-list[{}; {len}]",
                    DisplayDataType(field.data_type().clone())
                );
                return f.write_str(&s);
            }
            DataType::LargeList(field) => {
                let s = format!("large-list[{}]", DisplayDataType(field.data_type().clone()));
                return f.write_str(&s);
            }
            DataType::Struct(fields) => return write!(f, "struct[{}]", fields.len()),
            DataType::Union(fields, _, _) => return write!(f, "union[{}]", fields.len()),
            DataType::Map(field, _) => {
                return write!(f, "map[{}]", DisplayDataType(field.data_type().clone()))
            }
            DataType::Dictionary(_, _, _) => "dict",
            DataType::Decimal(_, _) => "decimal",
            DataType::Decimal256(_, _) => "decimal256",
            DataType::Extension(name, data_type, _) => {
                let s = format!("extension<{name}>[{}]", DisplayDataType(*data_type.clone()));
                return f.write_str(&s);
            }
        };
        f.write_str(s)
    }
}

/// Format `columns` into a [`Table`] using `names` as headers.
pub fn format_table<A, Ia, N, In>(columns: Ia, names: In) -> Table
where
    A: AsRef<dyn Array>,
    Ia: IntoIterator<Item = A>,
    N: AsRef<str>,
    In: IntoIterator<Item = N>,
{
    let mut table = Table::new();
    table.load_preset(presets::UTF8_FULL);

    const WIDTH_UPPER_BOUNDARY: u16 = 100;

    let names = names
        .into_iter()
        .map(|name| name.as_ref().to_owned())
        .collect::<Vec<_>>();
    let arrays = columns.into_iter().collect::<Vec<_>>();

    let (displayers, lengths): (Vec<_>, Vec<_>) = arrays
        .iter()
        .zip(names.iter())
        .map(|(array, name)| {
            let formatter = get_custom_display(name, array.as_ref(), "-");
            (formatter, array.as_ref().len())
        })
        .unzip();

    if displayers.is_empty() {
        return table;
    }

    let header = names
        .iter()
        .zip(arrays.iter().map(|array| array.as_ref().data_type()))
        .map(|(name, data_type)| {
            Cell::new(format!(
                "{}\n---\n{}",
                name.trim_start_matches("rerun.archetypes.")
                    .trim_start_matches("rerun.components.")
                    .trim_start_matches("rerun.datatypes.")
                    .trim_start_matches("rerun.controls.")
                    .trim_start_matches("rerun."),
                DisplayDataType(data_type.clone())
            ))
        });
    table.set_header(header);

    for row in 0..lengths[0] {
        let cells: Vec<_> = displayers
            .iter()
            .map(|disp| {
                let mut string = String::new();
                (disp)(&mut string, row).unwrap();
                let chars: Vec<_> = string.chars().collect();
                if chars.len() > WIDTH_UPPER_BOUNDARY as usize {
                    Cell::new(
                        chars
                            .into_iter()
                            .take(WIDTH_UPPER_BOUNDARY.saturating_sub(1).into())
                            .chain(['…'])
                            .collect::<String>(),
                    )
                } else {
                    Cell::new(string)
                }
            })
            .collect();
        table.add_row(cells);
    }

    table.set_content_arrangement(comfy_table::ContentArrangement::DynamicFullWidth);
    // NOTE: `Percentage` only works for terminals that report their sizes.
    let width = if table.width().is_some() {
        comfy_table::Width::Percentage((100.0 / arrays.len() as f32) as u16)
    } else {
        comfy_table::Width::Fixed(WIDTH_UPPER_BOUNDARY)
    };
    table.set_constraints(
        std::iter::repeat(comfy_table::ColumnConstraint::UpperBoundary(width)).take(arrays.len()),
    );

    table
}
