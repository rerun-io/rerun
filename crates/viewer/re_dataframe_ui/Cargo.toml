[package]
authors.workspace = true
description = "Rich table widget over DataFusion."
edition.workspace = true
homepage.workspace = true
license.workspace = true
name = "re_dataframe_ui"
publish = true
readme = "README.md"
repository.workspace = true
rust-version.workspace = true
version.workspace = true
include.workspace = true

[lints]
workspace = true

[package.metadata.docs.rs]
all-features = true

[dependencies]
re_arrow_util.workspace = true
re_chunk_store.workspace = true
re_dataframe.workspace = true
re_log_types.workspace = true
re_sorbet.workspace = true
re_tracing.workspace = true
re_types.workspace = true
re_types_core.workspace = true
re_ui.workspace = true
re_uri.workspace = true
re_viewer_context.workspace = true


ahash.workspace = true
arrow.workspace = true
datafusion.workspace = true
egui.workspace = true
egui_dnd.workspace = true
egui_table.workspace = true
nohash-hasher.workspace = true
parking_lot.workspace = true
serde.workspace = true
thiserror.workspace = true


[dev-dependencies]
static_assertions.workspace = true
