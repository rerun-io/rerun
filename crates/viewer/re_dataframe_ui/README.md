# re_dataframe_ui

Part of the [`rerun`](https://github.com/rerun-io/rerun) family of crates.

[![Latest version](https://img.shields.io/crates/v/re_dataframe_ui.svg)](https://crates.io/crates/re_dataframe_ui)
[![Documentation](https://docs.rs/re_dataframe_ui/badge.svg)](https://docs.rs/re_dataframe_ui)
![MIT](https://img.shields.io/badge/license-MIT-blue.svg)
![Apache](https://img.shields.io/badge/license-Apache-blue.svg)

Rich table widget over DataFusion.
