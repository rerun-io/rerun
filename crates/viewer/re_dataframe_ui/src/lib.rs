//! Rich table widget over `datafusion`.

mod datafusion_adapter;
mod datafusion_table_widget;
mod display_record_batch;
mod requested_object;
mod table_blueprint;
pub mod table_utils;

pub use datafusion_table_widget::DataFusionTableWidget;
pub use display_record_batch::{DisplayRecordBatch, DisplayRecordBatchError};
pub use requested_object::RequestedObject;
