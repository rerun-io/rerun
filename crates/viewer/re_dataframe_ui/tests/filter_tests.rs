use std::str::FromStr as _;
use std::sync::Arc;

use arrow::array::{
    Array, ArrayRef, BooleanArray, ListArray, ListBuilder, PrimitiveArray, PrimitiveBuilder,
    StringArray,
};
use arrow::buffer::{NullBuffer, OffsetBuffer};
use arrow::compute::cast;
use arrow::datatypes::{
    ArrowPrimitiveType, DataType, Field, Float32Type, Float64Type, Int8Type, Int16Type, Int32Type,
    Int64Type, Schema, TimeUnit, TimestampNanosecondType, UInt8Type, UInt16Type, UInt32Type,
    UInt64Type,
};
use arrow::record_batch::RecordBatch;
use datafusion::catalog::MemTable;
use datafusion::prelude::{DataFrame, SessionContext};
use jiff::ToSpan as _;

use re_dataframe_ui::{
    ComparisonOperator, Filter, FilterKind, FloatFilter, IntFilter, NonNullableBooleanFilter,
    Nullability, NullableBooleanFilter, TimestampFilter,
};
use re_viewer_context::external::tokio;

const COLUMN_NAME: &str = "column";
const SOME_TIMESTAMP: &str = "2025-09-23T11:47Z";

/// A single column of test data, with convenient constructors.
#[derive(Debug, Clone)]
struct TestColumn {
    field: Field,
    array: ArrayRef,
}

impl TestColumn {
    fn new(array: impl Array + 'static, nullable: bool) -> Self {
        let array = Arc::new(array) as ArrayRef;
        let field = Field::new(COLUMN_NAME, array.data_type().clone(), nullable);

        Self { array, field }
    }

    /// Create a primitive array with the provided data.
    fn primitive<T>(data: Vec<T::Native>) -> Self
    where
        T: ArrowPrimitiveType,
        PrimitiveArray<T>: From<Vec<T::Native>>,
    {
        Self::new(PrimitiveArray::<T>::from(data), false)
    }

    /// Create a nullable primitive array with the provided data.
    fn primitive_nulls<T>(data: Vec<Option<T::Native>>) -> Self
    where
        T: ArrowPrimitiveType,
        PrimitiveArray<T>: From<Vec<Option<T::Native>>>,
    {
        Self::new(PrimitiveArray::<T>::from(data), true)
    }

    /// Create a list array with the provided data and automatically injecting nulls as required.
    fn primitive_lists<T>(data: &[Vec<T::Native>], nullability: Nullability) -> Self
    where
        T: ArrowPrimitiveType,
        T::Native: Copy,
    {
        assert!(!data.is_empty());

        let value_builder = PrimitiveBuilder::<T>::new();
        let mut builder = ListBuilder::new(value_builder);

        for item in data {
            for inner_item in item {
                builder.values().append_value(*inner_item);
            }
            builder.append(true);
        }

        // inject outer null
        if nullability.outer {
            builder.append(false);
        }

        // inject inner nulls
        if nullability.inner {
            // lone null
            builder.values().append_null();
            builder.append(true);

            // find a non-empty inner item
            let non_empty_item = data
                .iter()
                .find(|item| !item.is_empty())
                .expect("there should be at least some non-empty data");

            // first null
            builder.values().append_null();
            for inner_item in non_empty_item {
                builder.values().append_value(*inner_item);
            }
            builder.append(true);

            // last null
            for inner_item in non_empty_item {
                builder.values().append_value(*inner_item);
            }
            builder.values().append_null();
            builder.append(true);
        }

        let array = builder.finish();

        Self::new(array, nullability.outer)
    }

    fn strings() -> Self {
        Self::new(
            StringArray::from(vec!["a", "b", "c", "ab", "A B", "aBc"]),
            false,
        )
    }

    fn strings_nulls() -> Self {
        Self::new(
            StringArray::from(vec![
                Some("a"),
                Some("b"),
                None,
                Some("ab"),
                Some("A B"),
                Some("aBc"),
            ]),
            true,
        )
    }

    fn strings_lists(nullability: Nullability) -> Self {
        // the primitive array stuff doesn't work for strings, so we go the manual way.
        let values = if nullability.inner {
            StringArray::from(vec![
                Some("a"),
                Some("b"),
                None,
                Some("ab"),
                None,
                Some("aBc"),
            ])
        } else {
            StringArray::from(vec!["a", "b", "c", "ab", "A B", "aBc"])
        };
        let offsets = OffsetBuffer::new(vec![0i32, 2, 4, 6].into());
        let strings_lists = ListArray::try_new(
            Arc::new(Field::new("item", DataType::Utf8, nullability.inner)),
            offsets,
            Arc::new(values),
            nullability
                .outer
                .then(|| NullBuffer::from(vec![true, false, true])),
        )
        .expect("failed to create a string list array");

        Self::new(strings_lists, nullability.outer)
    }

    fn bools() -> Self {
        Self::new(BooleanArray::from(vec![true, true, false]), false)
    }

    fn bools_nulls() -> Self {
        Self::new(
            BooleanArray::from(vec![Some(true), Some(true), None, Some(false)]),
            true,
        )
    }

    fn bool_lists(nullability: Nullability) -> Self {
        // the primitive array stuff doesn't work for bools, so we go the manual way.
        let values = if nullability.inner {
            BooleanArray::from(vec![
                Some(true),
                Some(false),
                None,
                Some(true),
                Some(false),
                None,
                Some(true),
                Some(false),
            ])
        } else {
            BooleanArray::from(vec![true, false, true, true, false, false, true, false])
        };

        let offsets = OffsetBuffer::new(vec![0i32, 1, 2, 4, 6, 8].into());
        let bool_lists = ListArray::try_new(
            Arc::new(Field::new("item", DataType::Boolean, nullability.inner)),
            offsets,
            Arc::new(values),
            nullability
                .outer
                .then(|| NullBuffer::from(vec![true, false, true, true, true])),
        )
        .expect("failed to create a bool list array");

        Self::new(bool_lists, nullability.outer)
    }

    fn timestamps(unit: TimeUnit) -> Self {
        let some_date = jiff::Timestamp::from_str(SOME_TIMESTAMP).expect("valid");

        let nano_column = Self::primitive::<TimestampNanosecondType>(vec![
            timestamp_to_nanos(some_date),
            timestamp_to_nanos(some_date - 1.hours()),
            timestamp_to_nanos(some_date - 24.hours()),
            timestamp_to_nanos(some_date - 8760.hours()),
            timestamp_to_nanos(some_date + 1.hours()),
            timestamp_to_nanos(some_date + 24.hours()),
            timestamp_to_nanos(some_date + 8760.hours()),
        ]);

        convert_timestamp_column(nano_column, unit)
    }

    fn timestamps_nulls(unit: TimeUnit) -> Self {
        let some_date = jiff::Timestamp::from_str(SOME_TIMESTAMP).expect("valid");

        let nano_column = Self::primitive_nulls::<TimestampNanosecondType>(vec![
            Some(timestamp_to_nanos(some_date)),
            Some(timestamp_to_nanos(some_date - 1.hours())),
            Some(timestamp_to_nanos(some_date - 24.hours())),
            Some(timestamp_to_nanos(some_date - 8760.hours())),
            None,
            Some(timestamp_to_nanos(some_date + 1.hours())),
            Some(timestamp_to_nanos(some_date + 24.hours())),
            None,
            Some(timestamp_to_nanos(some_date + 8760.hours())),
        ]);

        convert_timestamp_column(nano_column, unit)
    }

    fn timestamps_lists(unit: TimeUnit, nullability: Nullability) -> Self {
        let some_date = jiff::Timestamp::from_str(SOME_TIMESTAMP).expect("valid");

        let nano_column = Self::primitive_lists::<TimestampNanosecondType>(
            &[
                vec![],
                vec![timestamp_to_nanos(some_date)],
                vec![
                    timestamp_to_nanos(some_date - 1.hours()),
                    timestamp_to_nanos(some_date - 24.hours()),
                ],
                vec![
                    timestamp_to_nanos(some_date),
                    timestamp_to_nanos(some_date - 168.hours()),
                    timestamp_to_nanos(some_date - 8760.hours()),
                ],
                vec![
                    timestamp_to_nanos(some_date + 1.hours()),
                    timestamp_to_nanos(some_date + 24.hours()),
                ],
                vec![
                    timestamp_to_nanos(some_date),
                    timestamp_to_nanos(some_date + 168.hours()),
                    timestamp_to_nanos(some_date + 8760.hours()),
                ],
                vec![
                    timestamp_to_nanos(some_date),
                    timestamp_to_nanos(some_date - 8760.hours()),
                    timestamp_to_nanos(some_date + 8760.hours()),
                ],
            ],
            nullability,
        );

        convert_timestamp_column(nano_column, unit)
    }
}

fn timestamp_to_nanos(ts: jiff::Timestamp) -> i64 {
    ts.as_nanosecond()
        .try_into()
        .expect("timestamp is too large")
}

fn convert_timestamp_column(nano_column: TestColumn, unit: TimeUnit) -> TestColumn {
    if unit == TimeUnit::Nanosecond {
        nano_column
    } else {
        TestColumn::new(
            cast(nano_column.array.as_ref(), &DataType::Timestamp(unit, None))
                .expect("timestamp column cast failed"),
            nano_column.field.is_nullable(),
        )
    }
}

/// A temporary session context populated with a "test" dataframe and constructed from a bunch
/// of test columns.
struct TestSessionContext {
    ctx: SessionContext,
}

impl TestSessionContext {
    fn new(columns: impl IntoIterator<Item = TestColumn>) -> Self {
        let ctx = SessionContext::new();

        let (fields, arrays): (Vec<_>, Vec<_>) =
            columns.into_iter().map(|c| (c.field, c.array)).unzip();

        let schema = Arc::new(Schema::new_with_metadata(fields, Default::default()));

        ctx.register_table(
            "__test",
            Arc::new(
                MemTable::try_new(
                    Arc::clone(&schema),
                    vec![vec![
                        RecordBatch::try_new_with_options(schema, arrays, &Default::default())
                            .expect("failed to create the record batch"),
                    ]],
                )
                .expect("failed to create mem table"),
            ),
        )
        .expect("failed to register table");

        Self { ctx }
    }

    async fn df(&self) -> DataFrame {
        self.ctx
            .table("__test")
            .await
            .expect("test table not found")
    }

    async fn to_filtered_record_batch(&self, filter: &Filter) -> RecordBatch {
        let df = self.df().await;

        let schema = df.schema();
        let filter_expr = filter
            .as_filter_expression(schema)
            .expect("couldn't create an expression from filter");

        let mut record_batches = df
            .filter(filter_expr)
            .expect("failed to apply filter")
            .collect()
            .await
            .expect("failed to collect");

        assert_eq!(record_batches.len(), 1);

        record_batches
            .pop()
            .expect("we just checked that there is one record batch")
    }
}

#[derive(Debug)]
#[expect(dead_code)] // debug is excluded from dead code analysis
struct TestResult<'a> {
    op: FilterKind,
    field: Field,
    unfiltered: ArrayRef,
    filtered: &'a ArrayRef,
}

//note: this is a macro so insta is exposed to the actual test function
macro_rules! filter_snapshot {
    ($filter_op:expr, $test_column:expr, $case:expr) => {
        let filter = Filter::new(COLUMN_NAME, $filter_op.clone());
        let initial_field = $test_column.field.clone();
        let initial_column = $test_column.array.clone();

        let result = TestSessionContext::new([$test_column])
            .to_filtered_record_batch(&filter)
            .await;

        assert_eq!(result.columns().len(), 1);
        assert_eq!(&initial_field, result.schema().field(0));

        let final_column = result.column(0);


        let test_results = TestResult {
            op: $filter_op,
            field: initial_field,
            unfiltered: initial_column,
            filtered: final_column,
        };

        insta::with_settings!({
           snapshot_suffix => $case,
        },
        {
            insta::assert_debug_snapshot!(test_results);
        });
    };
}

#[tokio::test]
async fn test_int_compares() {
    let ints = TestColumn::primitive::<Int64Type>(vec![1, 2, 3, 4, 5]);
    let ints_nulls =
        TestColumn::primitive_nulls::<Int64Type>(vec![Some(1), Some(2), None, Some(4), Some(5)]);

    for op in ComparisonOperator::ALL {
        filter_snapshot!(
            FilterKind::Int(IntFilter::new(*op, Some(3))),
            ints.clone(),
            format!("{}_3", op.as_ascii())
        );

        filter_snapshot!(
            FilterKind::Int(IntFilter::new(*op, Some(4))),
            ints_nulls.clone(),
            format!("nulls_{}_4", op.as_ascii())
        );

        filter_snapshot!(
            FilterKind::Int(IntFilter::new(*op, None)),
            ints_nulls.clone(),
            format!("nulls_{}_unspecified", op.as_ascii())
        );
    }
}

/// Make sure we correctly handle all integer types.
#[tokio::test]
async fn test_int_all_types() {
    macro_rules! test_int_all_types_impl {
        ($ty:tt) => {
            filter_snapshot!(
                FilterKind::Int(IntFilter::new(ComparisonOperator::Eq, Some(3))),
                TestColumn::primitive::<$ty>(vec![1, 2, 3, 4, 5]),
                format!("{:?}", $ty {})
            )
        };
    }

    test_int_all_types_impl!(Int8Type);
    test_int_all_types_impl!(Int16Type);
    test_int_all_types_impl!(Int32Type);
    test_int_all_types_impl!(Int64Type);
    test_int_all_types_impl!(UInt8Type);
    test_int_all_types_impl!(UInt16Type);
    test_int_all_types_impl!(UInt32Type);
    test_int_all_types_impl!(UInt64Type);
}

#[tokio::test]
async fn test_int_lists() {
    let data = vec![
        vec![1, 2, 3],
        vec![2],
        vec![2, 2],
        vec![4, 5, 6],
        vec![7, 4, 9],
        vec![5, 2, 1],
    ];
    let int_lists = TestColumn::primitive_lists::<Int64Type>(&data, Nullability::NONE);
    let int_lists_nulls = TestColumn::primitive_lists::<Int64Type>(&data, Nullability::BOTH);

    for op in ComparisonOperator::ALL {
        filter_snapshot!(
            FilterKind::Int(IntFilter::new(*op, Some(2))),
            int_lists.clone(),
            format!("{}_2", op.as_ascii())
        );

        filter_snapshot!(
            FilterKind::Int(IntFilter::new(*op, Some(2))),
            int_lists_nulls.clone(),
            format!("nulls_{}_2", op.as_ascii())
        );
    }
}

#[tokio::test]
async fn test_float_compares() {
    let floats = TestColumn::primitive::<Float64Type>(vec![1.0, 2.0, 3.0, 4.0, 5.0]);
    let floats_nulls = TestColumn::primitive_nulls::<Float64Type>(vec![
        Some(1.0),
        Some(2.0),
        None,
        Some(4.0),
        Some(5.0),
    ]);

    for op in ComparisonOperator::ALL {
        filter_snapshot!(
            FilterKind::Float(FloatFilter::new(*op, Some(3.0))),
            floats.clone(),
            format!("{}_3.0", op.as_ascii())
        );

        filter_snapshot!(
            FilterKind::Float(FloatFilter::new(*op, Some(4.0))),
            floats_nulls.clone(),
            format!("nulls_{}_4", op.as_ascii())
        );
    }
}

/// Make sure we correctly handle all float types.
#[tokio::test]
async fn test_float_all_types() {
    macro_rules! test_float_all_types_impl {
        ($ty:tt) => {
            filter_snapshot!(
                FilterKind::Float(FloatFilter::new(ComparisonOperator::Eq, Some(3.0))),
                TestColumn::primitive::<$ty>(vec![1.0, 2.0, 3.0, 4.0, 5.0]),
                format!("{:?}", $ty {})
            )
        };
    }

    test_float_all_types_impl!(Float32Type);
    test_float_all_types_impl!(Float64Type);
}

#[tokio::test]
async fn test_float_lists() {
    let data = vec![
        vec![1.0, 2.0, 3.0],
        vec![2.0],
        vec![2.0, 2.0],
        vec![4.0, 5.0, 6.0],
        vec![7.0, 4.0, 9.0],
        vec![5.0, 2.0, 1.0],
    ];
    let float_lists = TestColumn::primitive_lists::<Float64Type>(&data, Nullability::NONE);
    let float_lists_nulls = TestColumn::primitive_lists::<Float64Type>(&data, Nullability::BOTH);

    for op in ComparisonOperator::ALL {
        filter_snapshot!(
            FilterKind::Float(FloatFilter::new(*op, Some(2.0))),
            float_lists.clone(),
            format!("{}_2.0", op.as_ascii())
        );

        filter_snapshot!(
            FilterKind::Float(FloatFilter::new(*op, Some(2.0))),
            float_lists_nulls.clone(),
            format!("nulls_{}_2.0", op.as_ascii())
        );
    }
}

#[tokio::test]
async fn test_string_contains() {
    filter_snapshot!(
        FilterKind::StringContains(String::new()),
        TestColumn::strings(),
        "empty"
    );

    filter_snapshot!(
        FilterKind::StringContains("a".to_owned()),
        TestColumn::strings(),
        "a"
    );

    filter_snapshot!(
        FilterKind::StringContains("a".to_owned()),
        TestColumn::strings(),
        "ab"
    );

    filter_snapshot!(
        FilterKind::StringContains("A".to_owned()),
        TestColumn::strings(),
        "a_uppercase"
    );

    filter_snapshot!(
        FilterKind::StringContains(String::new()),
        TestColumn::strings_nulls(),
        "nulls_empty"
    );

    filter_snapshot!(
        FilterKind::StringContains("a".to_owned()),
        TestColumn::strings_nulls(),
        "nulls_a"
    );
}

#[tokio::test]
async fn test_string_contains_list() {
    for &nullability in Nullability::ALL {
        filter_snapshot!(
            FilterKind::StringContains("ab".to_owned()),
            TestColumn::strings_lists(nullability),
            format!("{nullability:?}_ab")
        );
    }
}

/// Non-nullable filter should work regardless of nullability.
#[tokio::test]
async fn test_non_nullable_boolean_equals() {
    filter_snapshot!(
        FilterKind::NonNullableBoolean(NonNullableBooleanFilter::IsTrue),
        TestColumn::bools(),
        "true"
    );

    filter_snapshot!(
        FilterKind::NonNullableBoolean(NonNullableBooleanFilter::IsFalse),
        TestColumn::bools(),
        "false"
    );

    filter_snapshot!(
        FilterKind::NonNullableBoolean(NonNullableBooleanFilter::IsTrue),
        TestColumn::bools_nulls(),
        "nulls_true"
    );

    filter_snapshot!(
        FilterKind::NonNullableBoolean(NonNullableBooleanFilter::IsFalse),
        TestColumn::bools_nulls(),
        "nulls_false"
    );
}

#[tokio::test]
async fn test_nullable_boolean_equals() {
    filter_snapshot!(
        FilterKind::NullableBoolean(NullableBooleanFilter::IsTrue),
        TestColumn::bools_nulls(),
        "nulls_true"
    );

    filter_snapshot!(
        FilterKind::NullableBoolean(NullableBooleanFilter::IsFalse),
        TestColumn::bools_nulls(),
        "nulls_false"
    );

    filter_snapshot!(
        FilterKind::NullableBoolean(NullableBooleanFilter::IsNull),
        TestColumn::bools_nulls(),
        "nulls_null"
    );
}

#[tokio::test]
async fn test_boolean_equals_list_non_nullable() {
    for &nullability in Nullability::ALL {
        filter_snapshot!(
            FilterKind::NonNullableBoolean(NonNullableBooleanFilter::IsTrue),
            TestColumn::bool_lists(nullability),
            format!("{nullability:?}_true")
        );
    }
}

#[tokio::test]
async fn test_boolean_equals_list_nullable() {
    // Note: NullableBooleanFilter doesn't support Nullability::NONE, but that's ok because
    // NonNullableBooleanFilter is used in this case.
    for nullability in [Nullability::BOTH, Nullability::INNER, Nullability::OUTER] {
        filter_snapshot!(
            FilterKind::NullableBoolean(NullableBooleanFilter::IsNull),
            TestColumn::bool_lists(nullability),
            format!("{nullability:?}")
        );
    }
}

const ALL_TIME_UNITS: &[TimeUnit] = &[
    TimeUnit::Second,
    TimeUnit::Millisecond,
    TimeUnit::Microsecond,
    TimeUnit::Nanosecond,
];

#[tokio::test]
async fn test_timestamps() {
    // Note: this test intends to cover all column datatypes. It does not intend to cover all kinds
    // of timestamp filtering, which is already covered by unit tests.

    let some_date = jiff::Timestamp::from_str(SOME_TIMESTAMP).expect("valid");

    let all_filters = [
        (
            FilterKind::Timestamp(TimestampFilter::after(some_date)),
            "after",
        ),
        (
            FilterKind::Timestamp(TimestampFilter::after(some_date + 1.seconds())),
            "after_strict",
        ),
        (
            FilterKind::Timestamp(TimestampFilter::between(
                some_date - 168.hours(),
                some_date - 2.hours(),
            )),
            "between",
        ),
    ];

    for time_unit in ALL_TIME_UNITS {
        for (filter, case) in &all_filters {
            for nullable in [true, false] {
                filter_snapshot!(
                    filter.clone(),
                    if nullable {
                        TestColumn::timestamps_nulls(*time_unit)
                    } else {
                        TestColumn::timestamps(*time_unit)
                    },
                    format!(
                        "{case}_{time_unit:?}{}",
                        if nullable { "_nulls" } else { "" }
                    )
                );
            }
        }
    }
}

#[tokio::test]
async fn test_timestamps_list() {
    // Note: this test intends to cover all column datatypes. It does not intend to cover all kinds
    // of timestamp filtering, which is already covered by unit tests.

    let some_date = jiff::Timestamp::from_str(SOME_TIMESTAMP).expect("valid");

    let all_filters = [
        (
            FilterKind::Timestamp(TimestampFilter::after(some_date)),
            "after",
        ),
        (
            FilterKind::Timestamp(TimestampFilter::after(some_date + 1.seconds())),
            "after_strict",
        ),
        (
            FilterKind::Timestamp(TimestampFilter::between(
                some_date - 168.hours(),
                some_date - 2.hours(),
            )),
            "between",
        ),
    ];

    for (filter, case) in &all_filters {
        for &nullability in Nullability::ALL {
            filter_snapshot!(
                filter.clone(),
                TestColumn::timestamps_lists(TimeUnit::Nanosecond, nullability),
                format!("{case}_{nullability:?}")
            );
        }
    }
}
