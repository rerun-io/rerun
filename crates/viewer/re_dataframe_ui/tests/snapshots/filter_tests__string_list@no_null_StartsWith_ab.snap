---
source: crates/viewer/re_dataframe_ui/tests/filter_tests.rs
expression: test_results
---
TestResult {
    op: String(
        StringFilter {
            operator: StartsWith,
            query: "ab",
        },
    ),
    field: Field {
        name: "column",
        data_type: List(
            Field {
                name: "item",
                data_type: Utf8,
                nullable: false,
                dict_id: 0,
                dict_is_ordered: false,
                metadata: {},
            },
        ),
        nullable: false,
        dict_id: 0,
        dict_is_ordered: false,
        metadata: {},
    },
    unfiltered: ListArray
    [
      StringArray
    [
      "hello_ab",
      "a",
    ],
      StringArray
    [
      "b",
      "c",
    ],
      StringArray
    [
      "ab",
      "A B",
    ],
    ],
    filtered: ListArray
    [
      StringArray
    [
      "ab",
      "A B",
    ],
    ],
}
